sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -388,3 +391,62 @@ object ExtractSingleColumnNullAwareAntiJoin extends JoinSelectionHelper with Pre
     case _ => None
   }
 }
+
+/**
+ * An extractor for operations such as DELETE and MERGE that require rewriting data.
+ *
+ * This class extracts the following entities:
+ *  - the row-level command (such as [[DeleteFromTable]]);
+ *  - the read relation in the rewrite plan that can be either [[DataSourceV2Relation]] or
+ *  [[DataSourceV2ScanRelation]] depending on whether the planning has already happened;
+ *  - the current rewrite plan.
```

Comment:
It would be good to mention why this always finds the read relation rather than constructing the `RowLevelCommand` with a hard reference to it. My understanding is that it may be changed by the optimizer. It could be removed based on the condition and there may be more than one depending on the planning for UPDATE queries. Is that right?

Reply:
I kept the minimum required logic for group-based deletes for now. You are right, this extractor will change to support UPDATE and delta-based sources. What about updating the description once we make those changes? For now, there will be exactly one read relation.